### PR TITLE
add geneva generator

### PR DESCRIPTION
**Pull request by mkirsano**

### CODE REVIEW

**Comment by alberto-sanchez** [comment]
please test

File: geneva.spec [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+Source: git+https://stash.desy.de/scm/geneva/geneva-public.git?obj=master/%{realversion}&export=%{n}-%{realversion}&output=/%{n}-%{realversion}-%{tag}.tgz
+
+BuildRequires: cmake
+
+Requires: python py2-setuptools py2-numpy gsl boost lhapdf hepmc openloops pythia8
+
+%prep
+%setup -q -n %{n}-%{realversion}
+
+mkdir build
```
Comment (smuzaffar) [inline geneva.spec]: @mkirsano , can you please 
 - move the `%build` statement just above this line
 - create build directory outside of the source tree e.g something like https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_10_3_X/gcc700/root.spec#L35 . You also need to update cmake to use `../%{n}-%{realversion}`

File: geneva.spec [hunk 2]
```diff
@@ -0,0 +1,32 @@
+### RPM external geneva 1.0-RC3
+Source: git+https://stash.desy.de/scm/geneva/geneva-public.git?obj=master/%{realversion}&export=%{n}-%{realversion}&output=/%{n}-%{realversion}-%{tag}.tgz
+
+BuildRequires: cmake
```
Comment (smuzaffar) [inline geneva.spec]: add dependency on `gmake` here.

File: geneva.spec [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      -Dboost_ROOT=${BOOST_ROOT} \
+      -Dlhapdf_ROOT=${LHAPDF_ROOT} \
+      -Dhepmc_ROOT=${HEPMC_ROOT} \
+      -Dopenloops_ROOT=${OPENLOOPS_ROOT} \
+      -Dpythia8_ROOT=${PYTHIA8_ROOT} \
+      ..
+
+%build
+cd build
+make LHAPDF_DATA_PATH=${LHAPDF_ROOT}/share/LHAPDF
```
Comment (smuzaffar) [inline geneva.spec]: can this be build in parallel if yes then please add `%{makeprocesses}`

**Comment by mkirsano** [comment]
@smuzaffar, I did not understand about gamke, there is no such tool

**Comment by smuzaffar** [comment]
oops sorry typo, it is `gmake`

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
@mkirsano , it builds fine for python2 but fails for python3 [a]. 
[a]
```
.../external/gcc/7.0.0-omkpbe2/bin/c++ -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I.../external/geneva/1.0-RC3/build/include -I.../external/geneva/1.0-RC3/geneva-1.0-RC3/include -I.../external/python3/3.6.4-gnimlf/include/python3.6m -c .../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp -o .../external/geneva/1.0-RC3/build/python/build-temp.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.o -pedantic -Wall -Wextra -Wconversion -Wsign-conversion -Wdeprecated -Wno-unused-parameter -Wno-implicit-fallthrough -std=c++11
cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp: In function 'PyObject* Geneva::get_banner(PyObject*, PyObject*)':
.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp:37:11: error: 'PyString_FromString' was not declared in this scope
    return PyString_FromString(Geneva::header().c_str());
           ^~~~~~~~~~~~~~~~~~~
.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp:37:11: note: suggested alternative: 'PyLong_FromString'
    return PyString_FromString(Geneva::header().c_str());
           ^~~~~~~~~~~~~~~~~~~
           PyLong_FromString
.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp: In function 'PyObject* Geneva::get_options(PyObject*, PyObject*)':
.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp:51:53: error: 'PyString_FromString' was not declared in this scope
          PyDict_SetItemString(d, it->first.c_str(), PyString_FromString(it->second->getValueStr().c_str()));
                                                     ^~~~~~~~~~~~~~~~~~~
.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp:51:53: note: suggested alternative: 'PyLong_FromString'
          PyDict_SetItemString(d, it->first.c_str(), PyString_FromString(it->second->getValueStr().c_str()));
                                                     ^~~~~~~~~~~~~~~~~~~
                                                     PyLong_FromString
.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp: In function 'PyObject* Geneva::initcpp_extensions()':
.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp:141:4: error: 'Py_InitModule' was not declared in this scope
    Py_InitModule("cpp_extensions", cpp_extensions_methods);
    ^~~~~~~~~~~~~
.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp:141:4: note: suggested alternative: 'Py_Initialize'
    Py_InitModule("cpp_extensions", cpp_extensions_methods);
    ^~~~~~~~~~~~~
    Py_Initialize
.../external/geneva/1.0-RC3/geneva-1.0-RC3/python/extensions.cpp:142:1: warning: no return statement in function returning non-void [-Wreturn-type]
 }
```

**Comment by mkirsano** [comment]
What does it mean about python3? Is the merge possible?

**Comment by smuzaffar** [comment]
please test

**Comment by mkirsano** [comment]
I have to ask the authors about python3

**Comment by smuzaffar** [comment]
@mkirsano , yes merge it possible. It is just that this package will not be available for python3. Thanks for contacting developers.

**Comment by smuzaffar** [comment]
+externals
@fabiocos , this is new generator. Feel free to merge it for 10.4.X

**Comment by davidlange6** [comment]
"It is just that this package will not be available for python3. " 

that is to say the lifetime make be only though the 10_4_X cycle in production..

**Comment by fabiocos** [comment]
@davidlange6 @smuzaffar this is indeed a problem. This generator is under discussion since several years, and interesting from the physics point of view for precision studies. But if it is bound to python2 I agree we merge it now to kick it out in a few month: does it make sense? 

@mkirsano @efeyazgan @qliphy @alberto-sanchez did you contact the authors about this technical but important issue for the medium-long term maintainability?

**Comment by davidlange6** [comment]
> On Oct 12, 2018, at 9:31 AM, Fabio Cossutti <notifications@github.com> wrote:
> 
> @davidlange6 @smuzaffar this is indeed a problem. This generator is under discussion since several years, and interesting from the physics point of view for precision studies. But if it is bound to python2 I agree we merge it now to kick it out in a few month: does it make sense?

if its a matter of user analysis (eg, not cmsRun level) then I guess it will continue to function a while longer (unless a technical hurdle prevents that)

**Comment by fabiocos** [comment]
@davidlange6 well, I assume people will want to have official productions with it, with the idea of carrying it forward as we do for other tools (e.g. madgraph, powheg, sherpa...) So I am afraid it will not just be a Run2 standalone effort restricted to 10_4_X, at least this is not what I would expect

**Comment by davidlange6** [comment]
via the python api?

**Comment by bendavid** [comment]
Expected usage of this is similar to madgraph, with an external executable called from ExternalLHEProducer or similar producing an LHE file, which is then showered by pythia within CMSSW, so as long as python2 is available within the environment it should be ok.

**Comment by davidlange6** [comment]
Good - so lets get the answer about python 3 support (as end of life of python 2 is not so far away) and if not crazy merge this and move on..

**Comment by mkirsano** [comment]
The authors said that they don't have concrete plans about python 3 support, but if it is needed for CMS, thay can put some priority in this and work on it. So far they were responsive and helpful. In RC3 they solved all problems reported about RC2.

**Comment by davidlange6** [comment]
python 2 end of life is in 14 months. Supporting python 3 on that timescale would suit the current plan for CMS given what has been said about its usage and our expectation that we can support python2 and python3 together

**Comment by fabiocos** [comment]
@mkirsano @alberto-sanchez @efeyazgan @qliphy there was any clarification about the timescale for a python3 update of the existing code? The questions not "whether" we need it but "when" it might become available

**Comment by mkirsano** [comment]
I informed the authors that we need the compatibility with python 3 as soon as possible, in any case it should be compatible early next year. The answer is:

We'll look into migrating to python 3 and prioritize it accordingly.
I'll keep you updated ...

**Comment by smuzaffar** [comment]
please test
just to get a fresh test result

**Pull request merged**